### PR TITLE
Explode UUID map to tree directory with JSON files

### PR DESCRIPTION
**Pull request by eddowh**
Slight issue; when I write helper function to load tree directory + JSON files to nested dict, it will load with unescaped unicode characters (I tried so many ways to encode and decode with ascii and utf-8, still no help...)

Example:

This is `cn/beijing.json`:

```json
{
    "万寿西宫": "wanshouxigong",
    "东四": "dongsi",
    "农展馆": "nongzhanguan",
    "古城": "gucheng",
    "天坛": "tiantan",
    "奥体中心": "aotizhongxin",
    "官园": "guanyuan",
    "定陵": "dingling",
    "怀柔镇": "huairouzhen",
    "昌平镇": "changpingzhen",
    "海淀区万柳": "haidianqwanliu",
    "顺义新城": "shunyixincheng"
}
```

When I load this as part of a dictionary, however, it is like:

```python
{u'\u4e07\u5bff\u897f\u5bab': u'wanshouxigong',
 u'\u4e1c\u56db': u'dongsi',
 u'\u519c\u5c55\u9986': u'nongzhanguan',
 u'\u53e4\u57ce': u'gucheng',
 u'\u5929\u575b': u'tiantan',
 u'\u5965\u4f53\u4e2d\u5fc3': u'aotizhongxin',
 u'\u5b98\u56ed': u'guanyuan',
 u'\u5b9a\u9675': u'dingling',
 u'\u6000\u67d4\u9547': u'huairouzhen',
 u'\u660c\u5e73\u9547': u'changpingzhen',
 u'\u6d77\u6dc0\u533a\u4e07\u67f3': u'haidianqwanliu',
 u'\u987a\u4e49\u65b0\u57ce': u'shunyixincheng'}
```

### CODE REVIEW

File: openkongqi/utils.py [hunk 1]
```diff
@@ -15,3 +17,34 @@ def get_random_ua():
         ua_strings = json.load(fd)
 
     return random.choice(ua_strings)
+
+
+def tree_dir_json_to_nested_dict(base_dir):
```
Comment (hrbonz) [inline openkongqi/utils.py]: Let's replace that bit with [`os.walk`](https://docs.python.org/2/library/os.html#os.walk), we should also find a simpler name for that function.

**Comment by hrbonz** [comment]
I have this in my history about json and utf-8: http://stackoverflow.com/questions/9693699/python-json-and-unicode

**Comment by eddowh** [comment]
Here is an alternative function using `os.walk`. Personally, I think this function is harder to understand, less readable, and takes more lines of code.

Let me know if you want to go along with this implementation or the original recursive one.

```python
import os
import json

def get_directory_structure(basedir):
    """Create a nested dict given a tree directory structure.

    .. warning:: This function will return a dictionary that may contain
                 unicode characters that aren't escaped yet.

    :param base_dir: a valid directory or file path
    :returns: dict
    """
    dir = {}
    basedir = basedir.rstrip(os.sep)
    start = basedir.rfind(os.sep) + 1
    for path, dirs, files in os.walk(basedir):
    dir = {}
    rootdir = rootdir.rstrip(os.sep)
    start = rootdir.rfind(os.sep) + 1
    for path, dirs, files in os.walk(rootdir):
        folders = path[start:].split(os.sep)
        file_headers = map(lambda f: f[:-5], files)
        subdir = dict.fromkeys(file_headers)  # initializes each file key value with None
        # this loads the json data (pass if no valid JSON data)
        for header, file in zip(file_headers, files):
            file_path = os.path.join(path, file)
            try:
                subdir[header] = json.load(open(file_path))
            except ValueError as e:
                print e.message  # most likely == "No JSON object could be decoded"
                pass
        parent = reduce(dict.get, folders[:-1], dir)
        parent[folders[-1]] = subdir
    dir = dir[dir.keys()[0]]
    return dir
```

**Comment by hrbonz** [comment]
I changed the logic to a one dimension dict with UUIDs as keys, that makes everything more simple: https://gist.github.com/hrbonz/89c4f210fb565368a476

**Comment by hrbonz** [comment]
You need to look into the unicode issue with a solution from SO.

**Comment by eddowh** [comment]
The one dimensional solution key is feasible, but wouldn't it make more sense to make the `data/uuid` folder one directory with one-dimensional keys as well then? Then it's just a matter of reading from a directory and loading it without traversing nested directories with `os.walk` - that reduces complexity by quite a lot in my opinion.

However, if we're still considering the nested directories + dictionaries approach, I still stand by my original recursive solution.

That being said, we can also work with the one-dimensional keys if that is what you prefer. That also means that we will have to change the `get_uuid_map` function (or just get rid of it since it's simply access a value given a dictionary key)

**Comment by hrbonz** [comment]
The difference is not critical in my opinion, the nested folders helps organizing the data and the one dimension data structure is easier to manage. There's no reason to nest dict when loading the data.

File: openkongqi/utils.py [hunk 2]
```diff
@@ -15,3 +17,35 @@ def get_random_ua():
         ua_strings = json.load(fd)
 
     return random.choice(ua_strings)
+
+
+def load_sources(base_dir, sep):
+    """Create a one-dimensional dictionary given a tree directory sutrcutre.
```
Comment (hrbonz) [inline openkongqi/utils.py]: typo here

File: openkongqi/utils.py [hunk 3]
```diff
@@ -15,3 +17,35 @@ def get_random_ua():
...
+    for root, dirs, files in os.walk(base_dir):
+        relpath = os.path.relpath(root, base_dir)
+        if relpath != os.curdir:
+            reluuid = relpath.replace(os.sep, sep)
+        else:
+            reluuid = ''
+
+        for file in files:
+            filepath = os.path.join(root, file)
+            with open(filepath, 'r') as jsonfd:
```
Comment (hrbonz) [inline openkongqi/utils.py]: Do some `try...except` around the file opening and the json loading.

**Pull request merged**